### PR TITLE
Context provider

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

File: src/ContextEnricher/ContextEnricher.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+/**
+ * @psalm-type Backtrace = list<array{
+ *      file:string,
+ *      line:int,
+ *      function?:string,
+ *      class?:string,
+ *      type?:string,
+ *  }>
+ */
+final class ContextEnricher implements ContextEnricherInterface
```
Comment (samdark) [inline src/ContextEnricher/ContextEnricher.php:19]: Could it be named `ContextProvider`?

File: src/ContextEnricher/ContextEnricher.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    public function setExcludedTracePaths(array $excludedTracePaths): self
+    {
+        foreach ($excludedTracePaths as $excludedTracePath) {
+            /** @psalm-suppress DocblockTypeContradiction */
+            if (!is_string($excludedTracePath)) {
+                throw new InvalidArgumentException(
+                    sprintf(
+                        'The trace path must be a string, %s received.',
+                        gettype($excludedTracePath)
```
Comment (samdark) [inline src/ContextEnricher/ContextEnricher.php:83]: get_debug_type?

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,9 @@
 - Bug #98: Fix error on formatting trace, when it doesn't contain "file" and "line" (@vjik)
 - New #108: Support of nested values in message templates' variables, e. g. `{foo.bar}` (@vjik)
 - Bug #89: Fix error on parse messages, that contains variables that cannot cast to a string (@vjik)
+- New #109: Add context enricher (@vjik)
+- Chg #109: Deprecate `Logger` methods `setTraceLevel()` and `setExcludedTracePaths()` in favor of context enricher
```
Comment (samdark) [inline CHANGELOG.md:13]: Should README be adjusted?

File: src/Logger.php [hunk 4]
```diff
@@ -82,22 +73,21 @@ final class Logger implements LoggerInterface
...
      *
      * @param Target[] $targets The log targets.
+     * @param ContextEnricherInterface|null $contextEnricher The context enricher. If null, {@see ContextEnricher} with
+     * default parameters will be used.
      */
-    public function __construct(array $targets = [])
-    {
+    public function __construct(
+        array $targets = [],
+        ?ContextEnricherInterface $contextEnricher = null,
```
Comment (samdark) [inline src/Logger.php:87]: Should we provide a composite out of the box to be able to combine multiple context providers?

File: src/ContextEnricher/ContextEnricher.php (hunk 1 above)
Comment (rustamwin) [reply src/ContextEnricher/ContextEnricher.php:19]: Or `ContextProcessor`?

Comment (vjik) [reply src/ContextEnricher/ContextEnricher.php:19]: This class don't provide context, but add info to context. For example we can use several classes via composite enricher.

File: src/ContextEnricher/ContextEnricher.php (hunk 2 above)
Comment (vjik) [reply src/ContextEnricher/ContextEnricher.php:83]: Good idea. I will do it in separate PR (it need to chnage in several places).

File: CHANGELOG.md (hunk 3 above)
Comment (vjik) [reply CHANGELOG.md:13]: Currently README don't contain information about this methods.

File: src/Logger.php (hunk 4 above)
Comment (vjik) [reply src/Logger.php:87]: Added `CompositeContextEnricher`.

File: src/ContextEnricher/ContextEnricher.php (hunk 1 above)
Comment (vjik) [reply src/ContextEnricher/ContextEnricher.php:19]: If move merge logic into `Logger`, then "provider" is good... I will try to do this.

Comment (vjik) [reply src/ContextEnricher/ContextEnricher.php]: Done.

File: src/ContextProvider/CompositeContextProvider.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    ) {
+        $this->providers = $providers;
+    }
+
+    public function getContext(): array
+    {
+        $context = [];
+        foreach ($this->providers as $provider) {
+            $context = array_merge($context, $provider->getContext());
+        }
```
Comment (samdark) [inline src/ContextProvider/CompositeContextProvider.php:28]: See https://www.exakat.io/en/speeding-up-array_merge/

**Review by samdark — APPROVED** [review]

File: CHANGELOG.md (hunk 3 above)
Comment (samdark) [reply CHANGELOG.md]: Should it?

Comment (vjik) [reply CHANGELOG.md]: Yes. Bun I think it should be in separated section "Context providers".

Comment (vjik) [reply CHANGELOG.md]: Done

**Review by samdark — APPROVED** [review]

**Pull request merged**